Add Transaction Errors

s9latimm opened the pull request:
Add errors for transactions and handle them in parser / builder.

related https://github.com/se-sic/VaRA/issues/770
related https://github.com/se-sic/VaRA/issues/723

closes https://github.com/se-sic/VaRA/issues/744

s9latimm committed:
WIP: Update exec

s9latimm committed:
Apply suggestions from code review

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

vulder reviewed (commented):
from  the overall idea this looks fine
for debugging, maybe asan helps, could be that one function is return void and we just try to forward the return value -> UB


*later we just need to switch the bool with an enum that encode the error conditions

s9latimm committed:
Merge remote-tracking branch 'origin/vara-dev' into f-TransactionError

M	include/vara/Feature/FeatureModelTransaction.h
M	lib/Feature/FeatureModelParser.cpp
M	tools/fm-viewer/CMakeLists.txt
M	unittests/Feature/FeatureModelParser.cpp
M	unittests/resources/CMakeLists.txt
A	unittests/resources/xml/test_with_whitespaces.xml

s9latimm committed:
Add ErrorOr

A	include/vara/Feature/Error.h
M	include/vara/Feature/FeatureModel.h
M	include/vara/Feature/FeatureModelParser.h
M	include/vara/Feature/FeatureModelTransaction.h
M	lib/Feature/FeatureModelBuilder.cpp
M	lib/Feature/FeatureModelParser.cpp
M	lib/Feature/FeatureModelTransaction.cpp
M	unittests/Feature/FeatureModel.cpp
M	unittests/Feature/FeatureModelTransaction.cpp

s9latimm committed:
Fix error handling

M	include/vara/Feature/Error.h
M	include/vara/Feature/FeatureModel.h
M	include/vara/Feature/FeatureModelParser.h
M	include/vara/Feature/FeatureModelTransaction.h
M	lib/Feature/FeatureModelBuilder.cpp
M	lib/Feature/FeatureModelParser.cpp
M	unittests/Feature/CMakeLists.txt
M	unittests/Feature/FeatureModel.cpp
M	unittests/Feature/FeatureModelTransaction.cpp
M	unittests/Feature/NumericFeature.cpp

s9latimm committed:
Fix root handling and print

M	include/vara/Feature/Error.h
M	include/vara/Feature/FeatureModelTransaction.h
M	lib/Feature/FeatureModel.cpp
M	lib/Feature/FeatureModelBuilder.cpp
M	unittests/Feature/FeatureModelBuilder.cpp

s9latimm committed:
Failing SXFM parser tests

M	unittests/Feature/CMakeLists.txt

s9latimm committed:
Assert FM

M	unittests/Feature/FeatureModelSxfm.cpp

s9latimm committed:
Destroy builder

M	include/vara/Feature/FeatureModelBuilder.h
M	include/vara/Feature/FeatureModelParser.h
M	include/vara/Feature/FeatureModelTransaction.h
M	lib/Feature/FeatureModelParser.cpp

s9latimm committed:
Fix SXFM parser

M	include/vara/Feature/FeatureModelTransaction.h
M	lib/Feature/FeatureModelParser.cpp
M	unittests/Feature/FeatureModelSxfm.cpp
M	unittests/Feature/NumericFeature.cpp
M	unittests/resources/sxfm/test.sxfm

vulder commented on include/vara/Feature/Error.h:
```diff
@@ -0,0 +1,79 @@
+
+namespace vara::feature {
+
+enum ErrorCode {
```
```suggestion
enum FTErrorCode {
```
We should make it a bit clearer to what the ErrorCode actually belongs. So maybe `FTErrorCode` or `FeatureTransactionErrorCode` or `FeatureErrorCode`

vulder commented on include/vara/Feature/Error.h:
```diff
@@ -0,0 +1,79 @@
+
+template <typename ValueTy, typename ErrorTy = ErrorCode,
+          std::enable_if_t<!std::is_same_v<ValueTy, ErrorTy>, bool> = true>
+class ErrorOr {
```
Just use `llvm::ErrorOr`, we don't have to implement this here our selfs.
The only thing we need to do is make our errors std::error_code

vulder commented on include/vara/Feature/Error.h:
```diff
@@ -0,0 +1,79 @@
+  Result(ErrorCode Error) : ErrorOr<std::monostate, ErrorCode>(Error) {}
+
+  std::string toString() {
+    if (this->operator bool()) {
```
```suggestion
    if (*this) {
```

vulder commented on include/vara/Feature/Error.h:
```diff
@@ -0,0 +1,79 @@
+  std::variant<ValueTy, ErrorTy> Value;
+};
+
+class Result : public ErrorOr<std::monostate, ErrorCode> {
```
Hmm, this design looks a bit unnecessary complex. Why do we need a `Result` (+not a good name) type, when we just want to return an error code? An alternative could be to define the first state of the error code to `OK`, rename it to `****State` instead (or another name) of error and just always return that type.

Maybe, we could flip this on it's head and always return a `Foo<StatusCode, Value>` and have a special `Foo<StatusCode, void>` that represents a non return function. `Foo` then specifies that Value is only correct if `StatusCode` is `OK` (similar to an optional where we check if something is in there before we are allowed to access it).

s9latimm committed:
Add result

D	include/vara/Feature/Error.h
M	include/vara/Feature/FeatureModel.h
M	include/vara/Feature/FeatureModelParser.h
M	include/vara/Feature/FeatureModelTransaction.h
A	include/vara/Feature/Result.h
M	lib/Feature/FeatureModelBuilder.cpp
M	lib/Feature/FeatureModelParser.cpp
M	unittests/Feature/FeatureModel.cpp
M	unittests/Feature/FeatureModelTransaction.cpp

s9latimm committed:
Git is stupid

A	include/vara/Feature/Error.h

vulder committed:
Merge branch 'vara-dev' into f-TransactionError

vulder commented on include/vara/Feature/Result.h:
```diff
@@ -0,0 +1,115 @@
+} // namespace result
+
+template <typename ErrorTy, typename ValueTy = void>
+class Result {
```
Same the general `Result` imple looks like a utility to me, so Utils makes much more sense than putting it just in the Feature part.

vulder commented on include/vara/Feature/Result.h:
```diff
@@ -0,0 +1,115 @@
+
+namespace vara::feature {
+
+namespace result {
```
What's the idea behind putting this into it's own namespace?

vulder commented on include/vara/Feature/Result.h:
```diff
@@ -0,0 +1,115 @@
+};
+
+template <typename ErrorTy>
+class Error {
```
Why is the default `Error` class in the Result header? Shouldn't this be more in Utils?

vulder commented on include/vara/Feature/Result.h:
```diff
@@ -0,0 +1,115 @@
+template <typename ErrorTy>
+result::Error<ErrorTy> Error(ErrorTy E) {
+  return result::Error<ErrorTy>(std::move(E));
+}
```
This one looks nice (similar than rust :wink:).
But we should maybe additionally implement perfectly forwarding constructor calls with this pattern, so we can construct 'Foo' directly inside of the OK.

vulder commented on lib/Feature/FeatureModel.cpp:
```diff
@@ -93,7 +93,9 @@ std::unique_ptr<FeatureModel> FeatureModel::clone() const {
 
-  return FMB.buildFeatureModel();
+  auto FM = FMB.buildFeatureModel();
+  assert(FM);
```
Maybe a quick message, what happened here.

vulder commented on include/vara/Feature/Result.h:
```diff
@@ -0,0 +1,115 @@
+
+#include "llvm/Support/raw_ostream.h"
+
+#include "vara/Utils/VariantUtil.h"
```
```suggestion
#include "vara/Utils/VariantUtil.h"

#include "llvm/Support/raw_ostream.h"
```

vulder commented on include/vara/Feature/Result.h:
```diff
@@ -0,0 +1,115 @@
+public:
+  Ok(ValueTy V) : V(std::move(V)) {}
+
+  ValueTy operator*() { return std::move(V); }
```
I'm not sure if we should always move out the value, take a look at my proposal below, but also feel free to argu against it.

```suggestion
  ValueTy operator*() && { return extract_value(); }
  ValueTy operator*() & { return V; }
  
  ValueTy extract_value() { return std::move(V); }
```
In cases where we are an `rvalue` we can savely always move out the value, as it will not be needed afterwards.
If we are not, we "by default" copy the value out but offer and extra function `extract_value` to allow the caller to move out the value explicitly.

vulder commented on include/vara/Feature/Result.h:
```diff
@@ -0,0 +1,115 @@
+public:
+  Error(ErrorTy V) : E(std::move(V)) {}
+
+  ErrorTy operator*() { return std::move(E); }
```
Same here.

vulder commented on include/vara/Feature/Result.h:
```diff
@@ -0,0 +1,115 @@
+public:
+  Result(result::Ok<ValueTy> V) : Variant(std::move(V)) {}
+
+  Result(result::Error<ErrorTy> E) : Variant(std::move(E)) {}
```
Probably needs the same forwarding ctors as I mentioned below.

vulder commented on include/vara/Feature/Result.h:
```diff
@@ -0,0 +1,115 @@
+    return !std::holds_alternative<result::Error<ErrorTy>>(Variant);
+  }
+
+  [[nodiscard]] ValueTy operator*() {
```
Same case here for moving things out

vulder commented on include/vara/Feature/Error.h:
```diff
@@ -0,0 +1,74 @@
+  INCONSISTENT
+};
+
+namespace result {
```
I don't thing this should be in another namespace, just `vara:feature` as the Spezialisation is directly related to features.

vulder commented on include/vara/Feature/Error.h:
```diff
@@ -0,0 +1,74 @@
+  friend llvm::raw_ostream &operator<<(llvm::raw_ostream &OS,
+                                       const Error<FTErrorCode> &Error) {
+    switch (Error.E) {
+      // TODO(s9latimm): meaningful error msg
```
Open  TODO

vulder commented on include/vara/Feature/FeatureModelParser.h:
```diff
@@ -46,23 +46,23 @@ class FeatureModelXmlParser : public FeatureModelParser {
 
-  std::unique_ptr<FeatureModel> buildFeatureModel() override;
+  Result<FTErrorCode> verifyFeatureModel() override;
 
-  bool verifyFeatureModel() override;
+  std::unique_ptr<FeatureModel> buildFeatureModel() override;
```
Will we are at it, could you please add 1-2 sentences to comment the public API here. Now that we have docs, we should clean up missing comments as much as possible :laughing:

vulder commented on include/vara/Feature/FeatureModelTransaction.h:
```diff
@@ -296,12 +295,17 @@ class AddFeatureToModel : public FeatureModelModification {
-  void exec(FeatureModel &FM) override { (*this)(FM); }
+  Result<FTErrorCode> exec(FeatureModel &FM) override {
+    if (auto E = (*this)(FM); !E) {
+      return Error(E.getError());
```
Can we not just return `E` here?

vulder commented on include/vara/Feature/FeatureModelTransaction.h:
```diff
@@ -341,13 +350,13 @@ class RemoveFeatureFromModel : public FeatureModelModification {
       for (auto *C : F->getChildren<Feature>()) {
         // TODO (se-passau/VaRA#744): error, if call returns error (should not
-        // be possible)
+        //  be possible)
```
If this PR closes se-passau/VaRA#744 we should fix this here

vulder commented on include/vara/Feature/FeatureModelTransaction.h:
```diff
@@ -341,13 +350,13 @@ class RemoveFeatureFromModel : public FeatureModelModification {
     if (!F) {
       // TODO (se-passau/VaRA#744): error, does not exist
-      return;
+      return Error(MISSING_FEATURE);
```
I guess the comment above can be removed?

vulder commented on include/vara/Feature/FeatureModelTransaction.h:
```diff
@@ -356,9 +365,9 @@ class RemoveFeatureFromModel : public FeatureModelModification {
-        // TODO (se-passau/VaRA#744): error, non recursive 0remove on non leaf
-        // feature
-        return;
+        // TODO (se-passau/VaRA#744): error, non recursive remove on non leaf
+        //  feature
+        return Error(MISSING_FEATURE);
```
Is this correctly handled?

vulder commented on include/vara/Feature/FeatureModelTransaction.h:
```diff
@@ -442,7 +462,7 @@ class RemoveRelationshipFromModel : public FeatureModelModification {
     if (!P) {
       // TODO(se-passau/VaRA#744) handle this
-      return;
+      return Error(MISSING_PARENT);
```
Same here, should we remove the comment?

vulder commented on include/vara/Feature/FeatureModelTransaction.h:
```diff
@@ -628,9 +690,18 @@ class SetRoot : public FeatureModelModification {
-  RootFeature *operator()(FeatureModel &FM) {
+  Result<FTErrorCode, RootFeature *> operator()(FeatureModel &FM) {
+    if (FM.getRoot() && FM.getRoot()->getName() == Root->getName()) {
+      // TODO(s9latimm): swap root
```
Open TODO

vulder commented on include/vara/Feature/FeatureModelTransaction.h:
```diff
@@ -856,26 +946,36 @@ class FeatureModelModifyTransactionBase {
     }
-    return false;
-  };
+    abortImpl();
```
```suggestion

    abortImpl();
```

vulder reviewed (changes requested):
The Result/Error/Ok design looks nice.

We should also add unittests for the new utility classes Result/Error/Ok, which don't depend on anything Feature.

s9latimm committed:
Apply suggestions from code review

M	include/vara/Feature/Error.h
M	include/vara/Feature/FeatureModelTransaction.h
A	include/vara/Utils/Result.h
M	unittests/Utils/CMakeLists.txt
A	unittests/Utils/Result.cpp

s9latimm committed:
Merge remote-tracking branch 'origin/vara-dev' into f-TransactionError

M	.gitignore
M	bindings/python/tests/feature_model/test_feature_model.py
M	bindings/python/tests/writer/test_xml_writer.py
M	bindings/python/vara-feature/CMakeLists.txt
M	bindings/python/vara-feature/pybind_FeatureModel.cpp
D	bindings/python/vara-feature/pybind_FeatureModelXmlParser.cpp
M	bindings/python/vara-feature/pybind_Init.cpp
M	bindings/python/vara-feature/pybind_common.h
M	include/vara/Feature/FeatureModel.h
M	lib/Feature/FeatureModel.cpp
M	tools/fm-viewer/FeatureModelViewer.cpp

vulder commented on include/vara/Utils/Result.h:
```diff
@@ -0,0 +1,141 @@
+class Result {
+public:
+  Result(result::Ok<ValueTy> V)
+      : Variant(std::forward<result::Ok<ValueTy>>(V)) {}
```
What is the idea between forward here?
I guess you miss understood what I meant with my comment. I did not mean to use std::forward here on the Ok value (that would need a move) but to allow the Ok/Error values to construct the carried value by passing the correct parameters.

If you like, we can talk about this next Monday or so.

s9latimm replied:
Ah I see, probably we should talk about the remaining points anyway.

vulder committed:
Merge branch 'vara-dev' into f-TransactionError

s9latimm committed:
Apply suggestions from code review

M	include/vara/Feature/Error.h
M	include/vara/Feature/FeatureModelTransaction.h
M	include/vara/Utils/Result.h
M	lib/Feature/FeatureModel.cpp
M	unittests/Utils/Result.cpp

s9latimm committed:
Reformat

M	unittests/Utils/Result.cpp

s9latimm committed:
Forward construction

M	include/vara/Feature/Error.h
M	include/vara/Feature/FeatureModel.h
M	include/vara/Feature/FeatureModelTransaction.h
D	include/vara/Feature/Result.h
M	include/vara/Utils/Result.h
M	unittests/Utils/Result.cpp

vulder commented on include/vara/Feature/FeatureModelTransaction.h:
```diff
@@ -356,9 +364,9 @@ class RemoveFeatureFromModel : public FeatureModelModification {
-        // feature
-        return;
+        // TODO (se-passau/VaRA#744): error, non recursive remove on non leaf
+        //  feature
```
Reference to the closed issue se-sic/VaRA#744

vulder commented on include/vara/Feature/FeatureModelTransaction.h:
```diff
@@ -856,26 +944,37 @@ class FeatureModelModifyTransactionBase {
-        FM = nullptr;
-        return true;
+  Result<FTErrorCode> commitImpl() {
+    //    assert(FM && "Cannot commit Modifications without a FeatureModel
+    //    present.");
```
Out commented assert? Should this be included?

s9latimm replied:
This is not required anymore, as we already return an error if FM is not present, and may break with aborted transactions.

vulder commented on include/vara/Utils/Result.h:
```diff
@@ -0,0 +1,153 @@
+
+  Result(Error<ErrorTy> E) : Variant(std::move(E)) {}
+
+  template <typename OtherTy,
+            std::enable_if_t<!std::is_same_v<ValueTy, OtherTy>, bool> = true>
+  Result(Result<ErrorTy, OtherTy> Other) : Variant(Other.extract_error()) {}
```
What's the idea behind this impl? I would assume, if the error type is the same but the value type is different, we can take over the error. However, what I cannot figure out, what happens if `Other` does not contain an error, but a valid value of type `OtherTy`?

s9latimm replied:
With the current way of constructing, either an assertion or exception of `std::get: wrong index for variant` will be thrown, if `Other` does not actually contain an `Error`.
I did not come up with a better idea on how to infer that beforehand yet.

s9latimm committed:
Fix TODOs

M	include/vara/Feature/Error.h
M	include/vara/Feature/FeatureModelParser.h
M	include/vara/Feature/FeatureModelTransaction.h
M	lib/Feature/FeatureModelBuilder.cpp
M	unittests/Utils/Result.cpp

s9latimm committed:
Requested changes

M	include/vara/Feature/FeatureModelTransaction.h
M	include/vara/Utils/Result.h
M	unittests/Utils/Result.cpp
